libsql_client/dbapi2/types.py

```diff
@@ -0,0 +1,1423 @@
+    ":meta private:"
+    sql: str
+    tokens: Sequence[str]
+    is_dml: bool
+    __slots__ = ("sql", "tokens", "is_dml")
+
+    _dml_statements: ClassVar[Set[str]] = {
+        "INSERT", "UPDATE", "DELETE", "REPLACE",
+    }
+
+    @overload
+    def __init__(self, sql: str) -> None:
+        ...
+
+    @overload
+    def __init__(self, sql: str, _allow_instantiation: "Connection") -> None:
+        ...
+
+    def __init__(self, *args: object, **kwargs: object) -> None:
+        # NOTE do this to enable test_dbapi.py
+        # ModuleTests.test_disallow_instantiation(), it will call
+        # the connection to get a Statement, but the statement itself
+        # must not be instantiated directly.
+        # Test instantiates the result of: type(cx("select 1"))
+        con = kwargs.get("_allow_instantiation")
+        if not isinstance(con, Connection):
+            # matches check_disallow_instantiation() body
+            tp = self.__class__
+            mod = tp.__module__
+            name = tp.__name__
+            qualname = f"{mod}.{name}"
+            msg = f"cannot create '{qualname}' instances"
+            raise TypeError(msg)
+
+        assert len(args) == 1
+        sql = args[0]
+        if not isinstance(sql, str):
+            raise TypeError("sql must be a str")
+
+        if "\0" in sql:
+            raise ProgrammingError("the query contains a null character")
+        if sys.version_info[:2] >= (3, 11):
+            if len(sql) > con.getlimit(SQLITE_LIMIT_SQL_LENGTH):
+                raise DataError("query string is too large")
+
+        self.sql = sql
+        # see pysqlite_statement_create()
+        # https://github.com/python/cpython/blob/main/Modules/_sqlite/statement.c#L32
+        # however we need to parse the statement with iter_sql_statements()
+        # as we can't use sqlite3_prepare_v2()
+        itr = iter(iter_sql_statements(sql))
+        self.tokens = next(itr)
+        assert self.tokens
+        next_stmt = next(itr, None)
+        if next_stmt and next_stmt[0].upper() != "END":
+            msg = "You can only execute one statement at a time."
+            raise ProgrammingError(msg)
+        self.is_dml = self._check_is_dml(self.tokens)
+
+    @classmethod
+    def _check_is_dml(cls, tokens: Sequence[str]) -> bool:
+        # NOTE: this matches C pysqlite_statement_create()
+        # https://github.com/python/cpython/blob/main/Modules/_sqlite/statement.c#L75-L82
+        # but it's not fully correct as the language allows: WITH ...
+        # See https://www.sqlite.org/lang_insert.html
+        # https://www.sqlite.org/lang_update.html
+        # https://www.sqlite.org/lang_delete.html
+        # Syntax:
+        #    WITH RECURSIVE table_name ( col_name, other_col)
+        #    AS NOT MATERIALIZED ( SELECT ... )
+        return tokens[0].upper() in cls._dml_statements
+
+    @property
+    def is_readonly(self) -> bool:
+        # mimics sqlite3_stmt_readonly()
+        return not self.is_dml
```

Comment:
This does not mimic `sqlite3_stmt_readonly()`: for example, `CREATE TABLE` is not a readonly statement.

Reply:
do you have more that I should be handling? Or just "look at sqlite c code"?

but ok, needs fixing

Reply:
added ddl: "CREATE", "ALTER", "DROP", let me know if more is needed.